src/lib/optimizer/abstract_syntax_tree/abstract_ast_node.cpp

```diff
@@ -19,60 +19,100 @@ AbstractASTNode::AbstractASTNode(ASTNodeType node_type) : _type(node_type) {}
+  parent->set_child(child_side, nullptr);
+}
 
-void AbstractASTNode::clear_parent() {
-  // _parent is a weak_ptr that we need to lock
-  auto parent_ptr = parent();
-  if (!parent_ptr) return;
+void AbstractASTNode::clear_parents() {
+  // Don't use for-each loop here, as remove_parent manipulates the _parents vector
+  while (!_parents.empty()) {
+    auto parent = _parents.front().lock();
+    DebugAssert(parent, "Failed to lock parent");
+    remove_parent(parent);
+  }
+}
 
-  if (parent_ptr->_left_child.get() == this) {
-    parent_ptr->set_left_child(nullptr);
-  } else if (parent_ptr->_right_child.get() == this) {
-    parent_ptr->set_right_child(nullptr);
+ASTChildSide AbstractASTNode::get_child_side(const std::shared_ptr<AbstractASTNode>& parent) const {
+  if (parent->_children[0].get() == this) {
+    return ASTChildSide::Left;
+  } else if (parent->_children[1].get() == this) {
+    return ASTChildSide::Right;
   } else {
-    Fail("Invalid AST: ASTNode is not child of his parent.");
+    Fail("Specified parent node is not actually a parent node of this node.");
+    return ASTChildSide::Left;  // Make compilers happy
   }
-  _parent = {};
 }
 
-const std::shared_ptr<AbstractASTNode>& AbstractASTNode::left_child() const { return _left_child; }
+std::vector<ASTChildSide> AbstractASTNode::get_child_sides() const {
+  std::vector<ASTChildSide> child_sides;
+  child_sides.reserve(_parents.size());
 
-void AbstractASTNode::set_left_child(const std::shared_ptr<AbstractASTNode>& left) {
-  if (left == _left_child) return;
+  for (const auto& parent_weak_ptr : _parents) {
+    const auto parent = parent_weak_ptr.lock();
+    DebugAssert(parent, "Failed to lock parent");
+    child_sides.emplace_back(get_child_side(parent));
+  }
 
-  DebugAssert(left || !_right_child, "Node can't have right child and no left child");
+  return child_sides;
+}
 
-  _left_child = left;
-  if (left) left->_parent = shared_from_this();
+std::shared_ptr<AbstractASTNode> AbstractASTNode::left_child() const { return _children[0]; }
 
-  _on_child_changed();
+void AbstractASTNode::set_left_child(const std::shared_ptr<AbstractASTNode>& left) {
+  set_child(ASTChildSide::Left, left);
 }
 
-const std::shared_ptr<AbstractASTNode>& AbstractASTNode::right_child() const { return _right_child; }
+std::shared_ptr<AbstractASTNode> AbstractASTNode::right_child() const { return _children[1]; }
 
 void AbstractASTNode::set_right_child(const std::shared_ptr<AbstractASTNode>& right) {
-  if (right == _right_child) return;
-
-  DebugAssert(_left_child != nullptr, "Cannot set right child without having a left child.");
-
-  _right_child = right;
-  if (right) right->_parent = shared_from_this();
+  set_child(ASTChildSide::Right, right);
+}
 
-  _on_child_changed();
+std::shared_ptr<AbstractASTNode> AbstractASTNode::child(ASTChildSide side) const {
+  const auto child_index = static_cast<int>(side);
+  return _children[child_index];
```

Comment:
Also, this can lead to semantic ambiguities. Is `_children.size() == 1` equal to `left_child() && !right_child()`? The second one would ignore `nullptr`s.

Reply:
`_children` is a fixed size array, not a vector, for avoiding code duplication in `set_left/right_child()` an instead having one implementation in `set_child()`.